src/extension/index.ts

```diff
@@ -0,0 +1,193 @@
+    disposables.push(commands.registerCommand('sarif.showPanel', () => panel.show()));
+
+    // Diagnostics
+    const diagsAll = languages.createDiagnosticCollection('SARIF');
+    const setDiags = (doc: TextDocument) => {
+        if (doc.fileName.endsWith('.git')) return;
+        const artifactPath = baser.translateLocalToArtifact(doc.uri.path);
+        const severities = {
+            error: DiagnosticSeverity.Error,
+            warning: DiagnosticSeverity.Warning,
+        } as Record<string, DiagnosticSeverity>;
+        const diags = store.results
+            .filter(result => result._uri === artifactPath)
+            .map(result => new ResultDiagnostic(
+                regionToSelection(doc, result._region),
+                result._message ?? '—',
+                severities[result.level ?? ''] ?? DiagnosticSeverity.Information, // note, none, undefined.
+                result,
+            ));
+        diagsAll.set(doc.uri, diags);
+    };
+    workspace.textDocuments.forEach(setDiags);
+    workspace.onDidOpenTextDocument(setDiags);
+    workspace.onDidCloseTextDocument(doc => diagsAll.delete(doc.uri)); // Spurious *.git deletes don't hurt.
+    observe(store.logs, () => workspace.textDocuments.forEach(setDiags));
+
+    // Open Documents <-sync-> Store.logs
+    const syncActiveLog = async (doc: TextDocument) => {
+        if (!doc.fileName.match(/\.sarif$/i)) return;
+        if (store.logs.some(log => log._uri === doc.uri.toString())) return;
+        store.logs.push(...await loadLogs([doc.uri]));
+        panel.show();
+    };
+    workspace.textDocuments.forEach(syncActiveLog);
+    workspace.onDidOpenTextDocument(syncActiveLog);
+    workspace.onDidCloseTextDocument(doc => {
+        if (!doc.fileName.match(/\.sarif$/i)) return;
+        store.logs.removeWhere(log => log._uri === doc.uri.toString());
+    });
+
+    // Actions/Decorations for Call Trees
+    const decorationTypeCallout = window.createTextEditorDecorationType({
+        after: { color: new ThemeColor('problemsWarningIcon.foreground') }
+    });
+    const decorationTypeHighlight = window.createTextEditorDecorationType({
+        border: '1px',
+        borderStyle: 'solid',
+        borderColor: new ThemeColor('problemsWarningIcon.foreground'),
+    });
+    languages.registerCodeActionsProvider('*', {
+        provideCodeActions: (doc, _range, context) => {
+            if (context.only) return;
+
+            const diagnostic = context.diagnostics[0] as ResultDiagnostic | undefined;
+            if (!diagnostic) return;
+
+            const result = diagnostic?.result;
+            panel.select(result);
+            if (!result) return; // Don't clear the decorations until the next result is selected.
+
+            const editor = window.visibleTextEditors.find(editor => editor.document === doc);
+            if (!editor) return; // When would editor be undef?
+
+            const locations = result?.codeFlows?.[0]?.threadFlows?.[0]?.locations ?? [];
```

Comment:
>? [](start = 36, length = 1)

super nit: no need due to null-check above

Reply:
Good eyes! Fixed.

---
In reply to: [441204378](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441204378) [](ancestors = 441204378)